syntax = "proto3";

package raft;

service RaftService {
    // Election
    rpc RequestVote (VoteRequest) returns (VoteResponse);
    
    // Log Replication & Heartbeat
    rpc AppendEntries (AppendEntryRequest) returns (AppendEntryResponse);
    
    // New: Client Interaction
    rpc ExecuteCommand (CommandRequest) returns (CommandResponse);
}

message VoteRequest {
    int32 term = 1;
    string candidate_id = 2;
}

message VoteResponse {
    int32 term = 1;
    bool vote_granted = 2;
}

message LogEntry {
    int32 term = 1;
    string command = 2;
}

message AppendEntryRequest {
    int32 term = 1;
    string leader_id = 2;
    int32 prev_log_index = 3;
    int32 prev_log_term = 4;
    int32 leader_commit = 5;
    repeated LogEntry entries = 6; // The actual data being sent
}

message AppendEntryResponse {
    int32 term = 1;
    bool success = 2;
}

message CommandRequest {
    string command = 1;
}

message CommandResponse {
    bool success = 1;
    string message = 2;
    string leader_id = 3; // Tell client who the real leader is
}